File: src/main/java/me/centralhardware/znatoki/telegram/statistic/eav/types/Type.java
package me.centralhardware.znatoki.telegram.statistic.eav.types;

import io.vavr.control.Validation;
import org.telegram.telegrambots.meta.api.objects.Update;

import java.util.Objects;
import java.util.Optional;

public sealed interface Type permits Date, DateTime, Enumeration, Integer, Photo, Telephone, Text {

    String optionalText = "/skip для пропуска.";

    String format(String name, Boolean isOptional);
    Validation<String, Void> __validate(Update update, String...variants);

    Optional<String> extract(Update update);

    default Validation<String, Void> validate(Update update, String...variants){
        if (update.hasMessage() && Objects.equals(update.getMessage().getText(), "/skip")){
            return Validation.valid(null);
        }

        return __validate(update, variants);
    }

}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/eav/types/Type.kt
package me.centralhardware.znatoki.telegram.statistic.eav.types

import arrow.core.Either
import dev.inmo.tgbotapi.extensions.utils.extensions.raw.text
import dev.inmo.tgbotapi.types.message.abstracts.CommonMessage
import dev.inmo.tgbotapi.types.message.content.MessageContent
import dev.inmo.tgbotapi.types.message.content.TextContent
import kotlinx.serialization.Serializable

@Serializable
sealed interface Type {

    companion object {
        const val OPTIONAL_TEXT: String = "/skip для пропуска."
    }

    fun format(name: String, isOptional: Boolean): String

    fun validate(
        message: CommonMessage<MessageContent>,
        variants: List<String>,
    ): Either<String, Unit>

    fun extract(message: CommonMessage<MessageContent>): String? = message.text

    fun validate(
        message: CommonMessage<MessageContent>,
        vararg variants: String,
    ): Either<String, Unit> {
        if (message.content is TextContent && message.text == "/skip") {
            return Either.Right(Unit)
        }

        return validate(message, variants.toList())
    }

    fun name(): String = this.javaClass.simpleName
}

fun String.toType() =
    when (this) {
        "Date" -> Date
        "DateTime" -> DateTime
        "Enumeration" -> Enumeration
        "Integer" -> Integer
        "Photo" -> Photo
        "Telephone" -> Telephone
        "Text" -> Text
        else -> throw IllegalArgumentException()
    }
